samples: nrf9160: gps: update GPS sample to use the new GNSS API

tokangas opened the pull request:
Update GPS sample to use the new function based GNSS API.

Depends on https://github.com/nrfconnect/sdk-nrfxlib/pull/472.

Signed-off-by: Tommi Kangas <tommi.kangas@nordicsemi.no>

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-4430
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

tokangas commented:
Updated because of review changes to the GNSS API PR and rebased.

tokangas commented:
Rebased.

tokangas commented:
Updated manifest to point to modem lib PR https://github.com/nrfconnect/sdk-nrfxlib/pull/472.

tokangas commented:
README.rst needs to be also updated, that's still pending.

tokangas commented:
Updated README.rst.

tokangas commented:
Rebase.

tokangas commented:
The new GNSS API is now in NCS master, we should try to get this in as well so that we have a sample showing how it can be used.

jtguggedal commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -130,85 +133,97 @@ static int activate_lte(bool activate)
+
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		nmea_data = k_malloc(sizeof(struct nrf_modem_gnss_nmea_data_frame));
+
```
We should check that `nmea_data` isn't NULL before using it

jtguggedal commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -130,85 +133,97 @@ static int activate_lte(bool activate)
+					     sizeof(struct nrf_modem_gnss_nmea_data_frame),
+					     NRF_MODEM_GNSS_DATA_NMEA);
+		if (retval == 0) {
+			retval = k_msgq_put(&nmea_queue, &nmea_data, K_NO_WAIT);
```
It seems this return code is not checked. If that's intentional, it could be `void`ed explicitly

tokangas replied:
That's a bug, `nmea_data` needs to be freed if putting it to the message queue fails, I'll fix it.

jtguggedal commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -58,19 +57,23 @@ static const char *const at_commands[] = {
+static uint64_t fix_timestamp;
+static bool new_data;
+static bool gnss_blocked;
+static bool got_fix;
```
Should these flags be `volatile`, as they're changed in ISR and used in other threads?

tokangas replied:
Made these (and `agps_requested`) volatile.

tokangas replied:
During refactoring I removed most of these. Only `gnss_blocked` remains, that's now `volatile`.

lemrey commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -130,85 +133,97 @@ static int activate_lte(bool activate)
 	}
+}
+
+static int gnss_ctrl(uint32_t ctrl)
```
I'd say we could rework this function now that passing `GNSS_START` and `GNSS_STOP` is equivalent to calling the relative APIs, `nrf_modem_gnss_start()` and `nrf_modem_gnss_stop()`; we could simply call those and move the `GNSS_INIT_AND_START` code in the only place where it's called.

tokangas replied:
You are right, I will refactor this.

jtguggedal commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -130,85 +133,97 @@ static int activate_lte(bool activate)
+	switch (ctrl) {
+	case GNSS_INIT_AND_START:
+		retval = nrf_modem_gnss_init();
+		retval |= nrf_modem_gnss_event_handler_set(gnss_event_handler);
```
I think it would be beneficial to do error checks for each API call. I suspect this code will be copied around by users, so checking for each call will be easier to debug

tokangas replied:
Agree.

jtguggedal commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -130,85 +133,97 @@ static int activate_lte(bool activate)
-					sizeof(delete_mask));
 		if (retval != 0) {
-			printk("Failed to start GPS\n");
+			printk("Failed to initialize GNSS interface");
```
Missing `\n`

jtguggedal commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -130,85 +133,97 @@ static int activate_lte(bool activate)
+		break;
+
+	default:
+		return -1;
```
Could print a log message here that the control command was not recognized

tokangas replied:
Refactored the code as requested by @lemrey and removed the whole function.

lemrey reviewed (commented):
It would be good to add a few words (and/or a picture) in the conceptual documentation that explain what the Message Queue is used for, and how. This is a good place for users to see how to use the new interface and we could use it to explain the solution we came up with.

jtguggedal commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -410,40 +337,61 @@ int main(void)
 			print_nmea_data();
-			nmea_string_cnt = 0;
 		} else {
+			if (new_data) {
```
Could we use a semaphore or something instead for this synchronization so that we don't need to loop, but rather wait for new data? `k_poll` API can be used to wait for both sempahore and new data in message queue at the same time: https://docs.zephyrproject.org/latest/reference/kernel/other/polling.html

tokangas replied:
Added a new semaphore and use `k_poll()` to wait for either new PVT or NMEA data.

tokangas commented:
Thanks for your comments @lemrey and @jtguggedal! I addressed all other comments except @lemrey's suggestion to update documentation. I'll do that tomorrow morning.

I refactored the code quite a lot based on your comments. The biggest changes are that the `gnss_ctrl()` function is removed, a new semaphore and `k_poll()` is used for waiting new data and the usage of work to download the A-GPS data. Also GNSS is no longer stopped when A-GPS data is being downloaded, but only the PVT & NMEA output is temporarily suspended.

jtguggedal commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -128,324 +136,268 @@ static int activate_lte(bool activate)
+
+	activate_lte(true);
+	printk("Established LTE link\n");
+	if (open_supl_socket() == 0) {
```
I see that checkpatch doesn't mind, but personally I prefer blank line before `if` statements unless a variable is assigned on the line before and checked in the `if`

jtguggedal commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -128,324 +136,268 @@ static int activate_lte(bool activate)
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		nmea_data = k_malloc(sizeof(struct nrf_modem_gnss_nmea_data_frame));
+		if (nmea_data == NULL) {
+			break;
```
Would be usueful with a print here that memory allocation failed, indicating memory leak

tokangas replied:
Added the error print, although it is immediately overwritten because the sample clears the screen when it starts printing new PVT data. Anyway, it will still be visible in a log.

jtguggedal commented on samples/nrf9160/gps/src/main.c:
Candidate for blank line

jtguggedal commented on samples/nrf9160/gps/src/main.c:
Same here

jtguggedal reviewed (approved):
Looks good, thanks!

tokangas commented:
Modified the documentation as requested by @lemrey, addressed @jtguggedal's comments and added a check for return value when reading PVT data from the GNSS interface.

@peknis Can you check the documentation changes in my latest commit?

peknis commented on samples/nrf9160/gps/README.rst:
```diff
@@ -1,19 +1,26 @@
+The event handler function reads the received `NMEA`_ strings and puts those into the queue.
+The consumer loop reads from the queue and outputs the strings to the console.
+
+When SUPL support is enabled, a workqueue is used for downloading the `A-GPS`_ data.
```
```suggestion
When SUPL support is enabled, a work queue is used for downloading the `A-GPS`_ data.
```

tokangas replied:
I'm referring to Zephyr workqueues, which are written without a space: https://docs.zephyrproject.org/1.9.0/kernel/threads/workqueues.html

Should we add a link to the Zephyr documentation?

peknis replied:
Could be a good idea. Generally, work queue seems to be written as two separate words as I suggested, but Zephyr seems to disagree with that.

peknis commented on samples/nrf9160/gps/README.rst:
```diff
@@ -1,19 +1,26 @@
+The consumer loop reads from the queue and outputs the strings to the console.
+
+When SUPL support is enabled, a workqueue is used for downloading the `A-GPS`_ data.
+Downloading the data can take some time and using a workqueue ensures that the main thread is not blocked during the operation.
```
```suggestion
Downloading the data can take some time and using a work queue ensures that the main thread is not blocked during the operation.
```

tokangas replied:
Added link. Also added link to Zephyr message queue.

tokangas commented:
Final documentation tuning with @peknis's help.

peknis commented on samples/nrf9160/gps/README.rst:
```diff
@@ -1,19 +1,26 @@
-The sample first creates a `GNSS`_ socket.
-Then it reads the GPS data from the socket and parses the received frames to interpret the frame contents.
-See :ref:`nrfxlib:gnss_extension` for more information.
+The sample first initializes the `GNSS`_ interface.
```
Maybe leave this link out as we have it a few lines above.

peknis commented on samples/nrf9160/gps/README.rst:
```diff
@@ -1,19 +1,26 @@
+The event handler function reads the received NMEA strings and puts those into the message queue.
+The consumer loop reads from the queue and outputs the strings to the console.
+
+When SUPL support is enabled, a :ref:`Zephyr workqueue <zephyr:workqueues_v2>` is used for downloading the `A-GPS`_ data.
```
Leave out the A-GPS link as we have it already on line 11. Also consider getting rid of the same link under "Using the sample with..." just before the Testing section.

peknis commented on samples/nrf9160/gps/README.rst:
```diff
@@ -22,10 +29,10 @@ This sample operates in two different modes.
 
 In the default mode, the sample displays information from both PVT (Position, Velocity, and Time) and `NMEA`_ strings.
```
The first NMEA link is also pretty close to this, so consider leaving this out.
